### ISSUE TITLE
`datasets` needs to be installed to import the library 

### ISSUE THREAD

**Issue by fsndzomga** [outlines-dev/outlines#382]
### Describe the issue as clearly as possible:

I just started testing this library, and I get an error, even though I am just using the code examples from the README.
Am I doing something wrong ?

### Steps/code to reproduce the bug:

```python
import outlines.models as models
import outlines.text.generate as generate


model = models.transformers("gpt2-medium")

prompt = "Is 1+1=2? "
unguided = generate.continuation(model, max_tokens=30)(prompt)
guided = generate.regex(model, r"\s*([Yy]es|[Nn]o|[Nn]ever|[Aa]lways)", max_tokens=30)(
    prompt
)

print(unguided)
# Is 1+1=2?
#
# This is probably the most perplexing question.
# As I said in one of my articles describing how
# I call 2 and 1, there isn't

print(guided)
# Is 1+1=2? Always
```

### Expected result:

```shell
#
# This is probably the most perplexing question.
# As I said in one of my articles describing how
# I call 2 and 1, there isn't

# Is 1+1=2? Always
```

### Error message:

```shell
Traceback (most recent call last):
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/test-guidance.py", line 17, in <module>
    guided = generate.regex(model, r"\s*([Yy]es|[Nn]o|[Nn]ever|[Aa]lways)", max_tokens=30)(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/venv/lib/python3.11/site-packages/outlines/text/generate/regex.py", line 261, in regex
    return Regex(
           ^^^^^^
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/venv/lib/python3.11/site-packages/outlines/text/generate/regex.py", line 96, in __init__
    ) = create_fsm_index_tokenizer(regex_fsm, model.tokenizer)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/venv/lib/python3.11/site-packages/outlines/text/fsm.py", line 701, in create_fsm_index_tokenizer
    vocabulary, empty_token_ids = reduced_vocabulary(tokenizer)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/venv/lib/python3.11/site-packages/outlines/models/transformers.py", line 175, in __hash__
    from datasets.fingerprint import Hasher
ModuleNotFoundError: No module named 'datasets'
```

### Outlines/Python version information:

```python
Python 3.11.1 (v3.11.1:a7a450f84a, Dec  6 2022, 15:24:06) [Clang 13.0.0 (clang-1300.0.29.30)]
Traceback (most recent call last):
  File "/Users/fsndzomga/Documents/python_project/pinecone-chroma/test-guidance.py", line 33, in <module>
    import outlines; print("Outlines", outlines.__version__)
                                       ^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'outlines' has no attribute '__version__'
```

**Comment by rlouf** [comment]
Have you installed the `transformers` library?

**Comment by fsndzomga** [comment]
yes 

pip show transformers                          [🐍 system]
Name: transformers
Version: 4.33.3
Summary: State-of-the-art Machine Learning for JAX, PyTorch and TensorFlow
Home-page: https://github.com/huggingface/transformers
Author: The Hugging Face team (past and future) with the help of all our contributors (https://github.com/huggingface/transformers/graphs/contributors)
Author-email: transformers@huggingface.co
License: Apache 2.0 License
Location: /Users/fsndzomga/Documents/python_project/pinecone-chroma/venv/lib/python3.11/site-packages
Requires: filelock, huggingface-hub, numpy, packaging, pyyaml, regex, requests, safetensors, tokenizers, tqdm

**Comment by fsndzomga** [comment]
Thanks, installing datasets did the trick.

But now I get responses like this ??

" 
# Is 1+1=2?

unguided:  It should be named dead+2 for easy pluralization. Well, Bonus Larva Warbug online claimed it was Jurassic Era, while it actually
guided: Yes

"

**Comment by rlouf** [comment]
That looks correct to me. What is bothering you?

**Comment by fsndzomga** [comment]
Yes the guided worked in that example. I was just surprised by how weird the unguided can be.
Also I tried to run several examples and sometimes got this error:
```
Assertion failed: (isInt<33>(Addend) && "Invalid page reloc value."), function encodeAddend, file /Users/ci/miniconda3-arm64/envs/numba-builder/conda-bld/llvmdev_1684265191581/work/llvm/lib/ExecutionEngine/RuntimeDyld/Targets/RuntimeDyldMachOAArch64.h, line 210.
[1]    42716 abort      python test-outlines.py
```

Also I tried this example that I got from a hacker news thread and the guided response did not work, so I guess there are some limitations to how guidance one can set ?

```
import outlines.models as models
import outlines.text.generate as generate

model = models.transformers("gpt2-medium")
answer = generate.regex(model,
                        r"Rea[a-z']{,10} lik[a-z']{,10} acr[a-z']{,10} ene[a-z']{,10} sta[a-z']{,10}\.",
                        max_tokens=30)("A memorable phrase is:")

print(answer)
```

response: Rearissmation like acrowwoodandj enegumaintheb stavelamatica.

**Comment by fsndzomga** [comment]
Anyways, the idea behind the package is really nice. kudos to that.

**Comment by rlouf** [comment]
> I am also getting this error:
> 
> ```
> Assertion failed: (isInt<33>(Addend) && "Invalid page reloc value."), function encodeAddend, file /Users/ci/miniconda3-arm64/envs/numba-builder/conda-bld/llvmdev_1684265191581/work/llvm/lib/ExecutionEngine/RuntimeDyld/Targets/RuntimeDyldMachOAArch64.h, line 210.
> [1]    42716 abort      python test-outlines.py
> ```

Could you please give me more context so I can try to reproduce the error: the example you are trying to run, the `rng` seed, and the version of `numba` and `llvmlite` in your environment?

**Comment by lucidyan** [comment]
@rlouf error log with the Apple Silicon

```python
❯ python3 -m pip list | grep -E 'numba|llvmlite'
llvmlite                 0.41.0
numba                    0.58.0

❯ ipython3
Python 3.10.10 (main, May 26 2023, 00:45:17) [Clang 14.0.3 (clang-1403.0.22.14.1)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import outlines
   ...: model = outlines.models.transformers("mistralai/Mistral-7B-v0.1")
   ...:
   ...: prompt = """You are a sentiment-labelling assistant.
   ...: Is the following review positive or negative?
   ...:
   ...: Review: This restaurant is just awesome!
   ...: """
   ...: answer = outlines.generate.choice(model, ["Positive", "Negative"])(prompt)
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:05<00:00,  2.62s/it]
Assertion failed: (isInt<33>(Addend) && "Invalid page reloc value."), function encodeAddend, file /Users/ci/miniconda3-arm64/envs/numba-builder/conda-bld/llvmdev_1684265191581/work/llvm/lib/ExecutionEngine/RuntimeDyld/Targets/RuntimeDyldMachOAArch64.h, line 210.
[1]    39767 abort      ipython3
/Users/home/.pyenv/versions/3.10.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

**Comment by lucidyan** [comment]
Seems to be related with https://github.com/numba/numba/issues/8567

**Comment by gmarkall** [comment]
If you install the llvmite 0.42 and Numba 0.59 release candidates, this issue should no longer occur: https://numba.discourse.group/t/ann-numba-0-59-0rc1-and-llvmlite-0-42-0rc1/2329

**Comment by lucidyan** [comment]
@gmarkall Thank you! Could confirm that [updating](https://numba.discourse.group/t/ann-numba-0-59-0rc1-and-llvmlite-0-42-0rc1/2329) from `llvmlite==0.41.1` and `numba==0.58.1` to `lvmlite==0.42.0rc1` and `numba==0.59.0rc1` fix the problem on my Apple silicon Mac.

TL;DR do
`python3 -m pip install numba==0.59.0rc1 llvmlite==0.42.0rc1`

**Comment by RonanKMcGovern** [comment]
Still getting an error that datasets needs to be installed when running in Colab.

**Comment by cpfiffer** [comment]
For anyone who runs into this issue later, `datasets==2.21.0` worked for me where `2.20.0` did not.
